boards/shared/Src/Io/Io_SharedAdc.c

```diff
@@ -1,169 +1,140 @@
-    vrefint_index = vrefint_regular_rank - 1;
-    bool status   = true;
+    struct AdcInput *const adc_input = &adc_inputs[alloc_index];
 
-    if (HAL_ADCEx_Calibration_Start(hadc, ADC_SINGLE_ENDED) != HAL_OK)
-    {
-        status = false;
-        Error_Handler();
-    }
+    assert(alloc_index < MAX_NUM_OF_ADC);
+    assert(Io_GetAdcResolution(hadc, &adc_input->adc_max_value) == SUCCESS);
 
-    if (HAL_ADC_Start_DMA(hadc, &adc_values[0], NUM_ADC_CHANNELS) != HAL_OK)
-    {
-        status = false;
-        Error_Handler();
-    }
+    adc_input->hadc                = hadc;
+    adc_input->num_active_channels = hadc->Init.NbrOfConversion;
+    adc_input->adc_readings        = (uint16_t *)adc_readings[alloc_index];
+
+    HAL_ADCEx_Calibration_Start(hadc, ADC_SINGLE_ENDED);
+    HAL_ADC_Start_DMA(
+        hadc, (uint32_t *)adc_input->adc_readings,
+        adc_input->num_active_channels);
 
-    if (InitializeAdcMaxValue(hadc) != SUCCESS)
+    if (alloc_index == 0U)
     {
-        status = false;
-        Error_Handler();
+        // Initialize ADC instance containing VREFINT before initializing other
+        // ADCs
+        assert(vrefint_index <= adc_input->num_active_channels);
+
+        adc_inputs->vrefint_index = vrefint_index - 1;
+        // Store the pointer pointing to the measured VREFINT value
+        measured_vrefint_address =
+            &adc_input->adc_readings[adc_input->vrefint_index];
     }
-
-    // Only set adc_initialized to true if all initialization steps are
-    // error-free
-    if (status == true)
+    else
     {
-        adc_initialized = true;
+        UNUSED(vrefint_index);
     }
-}
-
-uint32_t Io_SharedAdc_GetAdcMaxValue(void)
-{
-    CheckAdcInitialized();
-
-    return (const uint32_t)adc_max_value;
-}
-
-const uint32_t *Io_SharedAdc_GetAdcValues(void)
-{
-    CheckAdcInitialized();
 
-    return (const uint32_t *)adc_values;
+    alloc_index++;
+    return adc_input;
 }
 
-float32_t Io_SharedAdc_GetActualVdda(void)
+float Io_SharedAdc_GetChannelVoltage(
+    const struct AdcInput *const adc_input,
+    size_t                       adc_rank)
 {
-    CheckAdcInitialized();
+    assert(adc_rank > 0 && adc_rank <= adc_input->num_active_channels);
 
-    // The actual VDDA voltage supplying the microcontroller can be calculated
-    // using the following formula:
+    // This function performs the following:
+    // (1) Calculate the microcontroller's actual VDDA value using the following
+    // formula
     //
-    //         3.3V x VREFINT_CAL
-    // VDDA = ---------------------
-    //           VREFINT_DATA
+    //                    3.3V x VREFINT_CAL
+    //  ACTUAL_VDDA = ---------------------------
+    //                      VREFINT_MEASURED
     //
-    // Where:
-    // - VREFINT_CAL is the VREFINT calibration value
-    // - VREFINT_DATA is the actual VREFINT output value converted by the ADC
+    // VREFINT_CAL is the VREFINT calibration value (stored at 0x1FFF7BA)
```

Comment:
``` suggestion
    // VREFINT_CAL is the VREFINT calibration value
```
In general we want to avoid duplicating constants, and in particular in comments where they can quickly become out of date.